File: aoc-2025/cpp/src/day06.cpp
#include <algorithm>
#include <cctype>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

const std::string DEFAULT_INPUT = "";
const std::string DEFAULT_OUTPUT = "";

bool is_col_empty(const std::vector<std::string> &lines, int col) {
  for (const auto &line : lines)
    if (col < (int)line.size() && line[col] != ' ')
      return false;
  return true;
}

long long part_one(const std::vector<std::string> &lines, int start, int end) {
  std::vector<std::string> tokens;

  for (const auto &line : lines) {
    if ((int)line.size() <= start)
      continue;
    int len = std::min((int)line.size(), end) - start;
    if (len <= 0)
      continue;

    std::string segment = line.substr(start, len);
    std::stringstream ss(segment);
    std::string token;
    while (ss >> token)
      tokens.push_back(token);
  }

  if (tokens.size() < 2)
    return 0;

  std::string opStr = tokens.back();
  tokens.pop_back();

  char op = opStr[0];
  if (op != '+' && op != '*')
    return 0;

  std::vector<long long> nums;
  for (const auto &t : tokens) {
    try {
      nums.push_back(std::stoll(t));
    } catch (...) {
    }
  }

  if (nums.empty())
    return 0;

  long long result = nums[0];
  for (size_t i = 1; i < nums.size(); ++i) {
    if (op == '+')
      result += nums[i];
    else
      result *= nums[i];
  }

  return result;
}

long long part_two(const std::vector<std::string> &lines, int start, int end) {
  std::vector<long long> nums;
  char op = 0;

  for (int x = end - 1; x >= start; --x) {
    std::string numStr;
    for (const auto &line : lines) {
      if (x < (int)line.size()) {
        char c = line[x];
        if (isdigit(c))
          numStr += c;
        else if (c == '+' || c == '*')
          op = c;
      }
    }
    if (!numStr.empty()) {
      try {
        nums.push_back(std::stoll(numStr));
      } catch (...) {
      }
    }
  }

  if (nums.empty() || (op != '+' && op != '*'))
    return 0;

  long long result = nums[0];
  for (size_t i = 1; i < nums.size(); ++i) {
    if (op == '+')
      result += nums[i];
    else
      result *= nums[i];
  }

  return result;
}

int main(int argc, char *argv[]) {
  std::string input_file = (argc > 1) ? argv[1] : DEFAULT_INPUT;
  std::cout << "Reading from: " << input_file << std::endl;

  std::ifstream file(input_file);
  if (!file.is_open()) {
    std::cerr << "Error opening file: " << input_file << std::endl;
    return -1;
  }

  std::vector<std::string> lines;
  std::string line;
  size_t max_width = 0;

  while (std::getline(file, line)) {
    lines.push_back(line);
    if (line.size() > max_width)
      max_width = line.size();
  }

  long long total_p1 = 0;
  long long total_p2 = 0;

  int start = -1;

  for (int x = 0; x <= (int)max_width; ++x) {
    bool empty = (x == (int)max_width) || is_col_empty(lines, x);

    if (!empty) {
      if (start == -1)
        start = x;
    } else {
      if (start != -1) {
        total_p1 += part_one(lines, start, x);
        total_p2 += part_two(lines, start, x);
        start = -1;
      }
    }
  }

  std::cout << "Part 1: " << total_p1 << std::endl;
  std::cout << "Part 2: " << total_p2 << std::endl;

  std::ofstream oFile(DEFAULT_OUTPUT);
  if (oFile.is_open())
    oFile << total_p1 << "\n" << total_p2 << "\n";

  return 0;
}

File: aoc-2025/rust/src/bin/day06.rs
//! https://adventofcode.com/2025/day/6
use std::env;
use std::fs;
use std::path::Path;

const DEFAULT_INPUT: &str = "";
const DEFAULT_OUTPUT: &str = "";

fn is_col_empty(grid: &[Vec<char>], col: usize) -> bool {
    grid.iter()
        .all(|row| row.get(col).copied().unwrap_or(' ') == ' ')
}

fn solve(content: &str, is_pt_two: bool) -> i64 {
    let lines: Vec<&str> = content.lines().collect();
    if lines.is_empty() {
        return 0;
    }

    let max_width = lines.iter().map(|l| l.len()).max().unwrap_or(0);

    let grid: Vec<Vec<char>> = lines
        .iter()
        .map(|l| {
            let mut chars: Vec<char> = l.chars().collect();
            while chars.len() < max_width {
                chars.push(' ');
            }
            chars
        })
        .collect();
    let mut ans: i64 = 0;
    let mut st = None;

    for x in 0..=max_width {
        let empty = if x == max_width {
            true
        } else {
            is_col_empty(&grid, x)
        };

        if !empty {
            if st.is_none() {
                st = Some(x);
            }
        } else if let Some(start) = st {
            if is_pt_two {
                ans += part_two(&grid, start, x);
            } else {
                ans += part_one(&grid, start, x);
            }
            st = None;
        }
    }
    ans
}

fn part_one(grid: &[Vec<char>], start: usize, end: usize) -> i64 {
    let mut tokens = Vec::new();
    for row in grid {
        let s: String = row[start..end].iter().collect();
        for part in s.split_whitespace() {
            tokens.push(part.to_string());
        }
    }

    if tokens.len() < 2 {
        return 0;
    }

    let op_str = tokens.pop().unwrap();
    let op = op_str.chars().next().unwrap_or(' ');

    let nums: Vec<i64> = tokens
        .iter()
        .filter_map(|t| t.parse::<i64>().ok())
        .collect();

    if nums.is_empty() {
        return 0;
    }

    let mut ans = nums[0];
    for &n in &nums[1..] {
        match op {
            '+' => ans += n,
            '*' => ans *= n,
            _ => {}
        }
    }
    ans
}

fn part_two(grid: &[Vec<char>], start: usize, end: usize) -> i64 {
    let mut nums = Vec::new();
    let mut op = ' ';

    for col in (start..end).rev() {
        let mut s = String::new();
        for row in grid {
            let c = row[col];
            if c.is_ascii_digit() {
                s.push(c);
            } else if c == '+' || c == '*' {
                op = c;
            }
        }

        if !s.is_empty() {
            if let Ok(n) = s.parse::<i64>() {
                nums.push(n);
            }
        }
    }

    if nums.is_empty() || (op != '+' && op != '*') {
        return 0;
    }

    let mut ans = nums[0];
    for &n in &nums[1..] {
        match op {
            '+' => ans += n,
            '*' => ans *= n,
            _ => {}
        }
    }
    ans
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let input_path = args.get(1).map(|s| s.as_str()).unwrap_or(DEFAULT_INPUT);
    println!("Reading from: {}", input_path);

    let content = fs::read_to_string(input_path).expect("Could not read input file");

    let p1 = solve(&content, false);
    let p2 = solve(&content, true);

    println!("Part 1: {}", p1);
    println!("Part 2: {}", p2);

    if let Some(parent) = Path::new(DEFAULT_OUTPUT).parent() {
        let _ = fs::create_dir_all(parent);
    }
    let _ = fs::write(DEFAULT_OUTPUT, format!("{}\n{}\n", p1, p2));
}
